sklearn/preprocessing/_encoders.py

```diff
@@ -264,233 +282,106 @@ class OneHotEncoder(_BaseEncoder):
       iterables and a multilabel format, e.g. a (samples x classes) binary
       matrix indicating the presence of a class label.
     """
-
     def __init__(self, categories='auto', drop=None, sparse=True,
                  dtype=np.float64, handle_unknown='error'):
+
         self.categories = categories
         self.sparse = sparse
         self.dtype = dtype
         self.handle_unknown = handle_unknown
         self.drop = drop
 
-    def _validate_keywords(self):
-        if self.handle_unknown not in ('error', 'ignore'):
-            msg = ("handle_unknown should be either 'error' or 'ignore', "
-                   "got {0}.".format(self.handle_unknown))
+    @property
+    def drop_idx_(self):
+        return np.array([encoder.drop_idx_ for encoder in self._encoders],
+                        dtype=np.int_)
+
+    def _fit(self, X):
+        """Validate keywords and fit `X` and return `X_list`."""
+        self._validate_keywords()
+        X_list = self._check_X(X)
+        n_features = len(X_list)
+
+        categories = self._check_categories(n_features)
+        drop_kwargs = self._check_drop(n_features)
+
+        encoders = [
```

Comment:
the pattern

```python
encoders = [...]
self._fit_list(...)
```

will be common to each encoder. Would it make sense to have an abstract method across encoder which should be defined in all children?

Then `_fit` will just be an input validation (which could also be an abstract method) and the above pattern (+ additinonal thing depending of the encoder)


```python
def _fit(self, X):
    X = self._validate_input(X)
    self._generate_and_fit_encoder(X)
    ...
```

Reply:
But be aware that this is not clear to me :)

Reply:
I am glad you noticed this! One of my designs had something similar where `_fit` was an abstract method. It was even possible to get to a point where the child class only needed to define `_fit` and `_EncoderUnion` can define `fit`, `transform`, and `fit_transform`.  It works, but it felt a little too magically. (There were additional lines to overwrite the docstring for `transform`, and `fit_transform` for a specific child encoder. For example, `OneHotEncoder.transform` can output sparse matrices.)

The benefits of this PR, is that `OrdinalEncoder.fit` explicitly defines the encoders, and explicitly passes it up into the parent class to process.